flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/FileSplitReadOperator.java

```diff
@@ -0,0 +1,176 @@
+	private transient TimestampedCollector<OUT> collector;
+
+	private Configuration configuration;
+	private FileInputFormat<OUT> format;
+	private TypeInformation<OUT> typeInfo;
+	private transient TypeSerializer<OUT> serializer;
+
+	public FileSplitReadOperator(FileInputFormat<OUT> format, TypeInformation<OUT> typeInfo, Configuration configuration) {
+		this.format = format;
+		this.typeInfo = typeInfo;
+		this.configuration = configuration;
+		this.unReadSplits = new ConcurrentLinkedQueue<>();
+
+		// this is for the extra thread that is reading,
+		// the tests were not terminating because the success
+		// exception was caught by the extra thread, and never
+		// forwarded.
+
+		// for now, and to keep the extra thread to avoid backpressure,
+		// we just disable chaining for the operator, so that it runs on
+		// another thread. This will change later for a cleaner design.
+		setChainingStrategy(ChainingStrategy.NEVER);
+	}
+
+	@Override
+	public void open() throws Exception {
+		super.open();
+
+		this.format.configure(configuration);
+		this.collector = new TimestampedCollector<OUT>(output);
+		this.checkpointLock = getContainingTask()
+			.getCheckpointLock();
+		this.serializer = typeInfo.createSerializer(getRuntimeContext().getExecutionConfig());
+
+		this.reader = new SplitReader(unReadSplits, format, serializer, collector, checkpointLock);
+		this.reader.start();
+	}
+
+	@Override
+	public void processElement(StreamRecord<FileInputSplit> element) throws Exception {
+		LOG.info("Reading Split: " + element.getValue());
+		this.unReadSplits.add(element.getValue());
+	}
+
+	@Override
+	public void processWatermark(Watermark mark) throws Exception {
+		output.emitWatermark(mark);
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+		this.reader.cancel();
+		this.collector.close();
+		this.format.close();
+	}
+
+	private class SplitReader<OUT> extends Thread {
+
+		private boolean isRunning;
+
+		private final Queue<FileInputSplit> splits;
+
+		private final FileInputFormat<OUT> format;
+		private final TypeSerializer<OUT> serializer;
+
+		private final Object checkpointLock;
+		private final TimestampedCollector<OUT> collector;
+
+		SplitReader(Queue<FileInputSplit> splits,
+					FileInputFormat<OUT> format,
+					TypeSerializer<OUT> serializer,
+					TimestampedCollector<OUT> collector,
+					Object checkpointLock) {
+			this.format = checkNotNull(format, "Unspecified FileInputFormat.");
+			this.serializer = checkNotNull(serializer, "Unspecified Serialized.");
+
+			this.splits = splits;
+			this.collector = collector;
+			this.checkpointLock = checkpointLock;
+			this.isRunning = true;
+		}
+
+		@Override
+		public void run() {
+			FileInputSplit split = null;
+
+			try {
+				while(this.isRunning) {
+
+					// get the next split to read
+					split = this.splits.poll();
+					if(split == null) {
+						Thread.sleep(50);
+						continue;
+					}
+
+					this.format.open(split);
```

Comment:
This should be: 
```
boolean isOpen = false;
try {
    this.format.open(split);
    isOpen = true;
    OUT nextElement = serializer.createInstance();
    do {
        nextElement = format.nextRecord(nextElement);
        if(nextElement != null) {
            synchronized (checkpointLock) {
                collector.collect(nextElement);
                // checkpointing should be here.
            }
        }
    } while(nextElement != null && !format.reachedEnd());
} finally {
    if (isOpen) {
        this.format.close();
    }
}
```
This is somewhat complicated and nonintuitive but I think we're one the save site like this... 😅